src/exo/cursors.py

```diff
@@ -0,0 +1,591 @@
+        return weakref.ref(n)
+
+    # ------------------------------------------------------------------------ #
+    # Navigation (implementation)
+    # ------------------------------------------------------------------------ #
+
+    def parent(self) -> Node:
+        if not self._path:
+            raise InvalidCursorError('cursor does not have a parent')
+        return Node(self._proc, self._path[:-1])
+
+    def before(self, dist=1) -> Gap:
+        return self._translate(Gap, self._path, 1 - dist)
+
+    def after(self, dist=1) -> Gap:
+        return self._translate(Gap, self._path, dist)
+
+    def prev(self, dist=1) -> Node:
+        return self._translate(Node, self._path, -dist)
+
+    def next(self, dist=1) -> Node:
+        return self._translate(Node, self._path, dist)
+
+    # ------------------------------------------------------------------------ #
+    # Navigation (children)
+    # ------------------------------------------------------------------------ #
+
+    def child(self, attr, i=None) -> Node:
+        _node = getattr(self._node(), attr)
+        if i is not None:
+            _node = _node[i]
+        elif isinstance(_node, list):
+            raise ValueError('must index into block attribute')
+        cur = Node(self._proc, self._path + [(attr, i)])
+        # noinspection PyPropertyAccess
+        # cached_property is settable, bug in static analysis
+        cur._node_ref = weakref.ref(_node)
+        return cur
+
+    def children(self) -> Iterable[Node]:
+        n = self._node()
+        # Top-level proc
+        if isinstance(n, LoopIR.proc):
+            yield from self._children_from_attrs(n, 'body')
+        # Statements
+        elif isinstance(n, (LoopIR.Assign, LoopIR.Reduce)):
+            yield from self._children_from_attrs(n, 'idx', 'rhs')
+        elif isinstance(n, (LoopIR.WriteConfig, LoopIR.WindowStmt)):
+            yield from self._children_from_attrs(n, 'rhs')
+        elif isinstance(n, (LoopIR.Pass, LoopIR.Alloc, LoopIR.Free)):
+            yield from []
+        elif isinstance(n, LoopIR.If):
+            yield from self._children_from_attrs(n, 'cond', 'body', 'orelse')
+        elif isinstance(n, (LoopIR.ForAll, LoopIR.Seq)):
+            yield from self._children_from_attrs(n, 'hi', 'body')
+        elif isinstance(n, LoopIR.Call):
+            yield from self._children_from_attrs(n, 'args')
+        # Expressions
+        elif isinstance(n, LoopIR.Read):
+            yield from self._children_from_attrs(n, 'idx')
+        elif isinstance(n, (LoopIR.Const, LoopIR.WindowExpr, LoopIR.StrideExpr,
+                            LoopIR.ReadConfig)):
+            yield from []
+        elif isinstance(n, LoopIR.USub):
+            yield from self._children_from_attrs(n, 'arg')
+        elif isinstance(n, LoopIR.BinOp):
+            yield from self._children_from_attrs(n, 'lhs', 'rhs')
+        elif isinstance(n, LoopIR.BuiltIn):
+            yield from self._children_from_attrs(n, 'args')
```

Comment:
There should be an `else: assert False, f"case {type(n)} unsupported"` or something like that

Reply:
Fixed.